snuba/util.py

```diff
@@ -102,64 +167,39 @@ def column_expr(column_name, body, alias=None, aggregate=None):
         expr = escape_col(column_name)
 
     if aggregate:
-        if expr:
-            expr = u'{}({})'.format(aggregate, expr)
-            if aggregate == 'uniq':  # default uniq() result to 0, not null
-                expr = 'ifNull({}, 0)'.format(expr)
-        else:  # This is the "count()" case where the '()' is already provided
-            expr = aggregate
+        expr = function_expr(aggregate, expr)
 
     alias = escape_col(alias or column_name)
 
     return alias_expr(expr, alias, body)
 
 
 def complex_column_expr(expr, body, depth=0):
-    # TODO instead of the mutual recursion between column_expr and complex_column_expr
-    # we should probably encapsulate all this logic in a single recursive column_expr
-    if depth == 0:
-        # we know the first item is a function
-        ret = expr[0]
-        assert SAFE_FUNCTION_RE.match(ret)
-        expr = expr[1:]
-
-        # if the last item of the toplevel is a string, it's an alias
-        alias = None
-        if len(expr) > 1 and isinstance(expr[-1], six.string_types):
-            alias = expr[-1]
-            expr = expr[:-1]
-    else:
-        # is this a nested function call?
-        if len(expr) > 1 and isinstance(expr[1], tuple):
-            ret = expr[0]
-            assert SAFE_FUNCTION_RE.match(ret)
-            expr = expr[1:]
+    function_tuple = is_function(expr, depth)
+    if function_tuple is None:
+        raise ValueError('complex_column_expr was given an expr %s that is not a function at depth %d.' % (expr, depth))
+
+    name, args, alias = function_tuple
+    out = []
+    i = 0
+    while i < len(args):
+        next_2 = args[i:i+2]
+        if is_function(next_2, depth+1):
+            out.append(complex_column_expr(next_2, body, depth+1))
+            i += 2
         else:
-            ret = ''
-
-    # emptyIfNull(col) is a simple pseudo function supported by Snuba that expands
-    # to the actual clickhouse function ifNull(col, '') Until we figure out the best
-    # way to disambiguate column names from string literals in complex functions.
-    if ret == 'emptyIfNull' and len(expr) >= 1 and isinstance(expr[0], tuple):
-        ret = 'ifNull'
-        expr = (expr[0] + ("''",),) + expr[1:]
-
-    first = True
-    for subexpr in expr:
-        if isinstance(subexpr, tuple):
-            ret += '(' + complex_column_expr(subexpr, body, depth + 1) + ')'
-        else:
-            if not first:
-                ret += ', '
-            if isinstance(subexpr, six.string_types):
-                ret += column_expr(subexpr, body)
+            nxt = args[i]
+            if is_function(nxt, depth + 1):  # Embedded function
```

Comment:
I guess maybe add a comment that this extra level of function nesting isn't required and should be removed once the queries issued by the sentry side have been changed.